colossalai/nn/parallel.py

```diff
@@ -76,3 +78,54 @@ def zero_grad(self, set_to_none: bool = False) -> None:
                     else:
                         p._saved_grad.requires_grad_(False)
                     p._saved_grad.zero_()
+
+
+class ColoDDPV2(ColoDDP):
+
+    def __init__(self, module: torch.nn.Module, chunk_manager: ChunkManager) -> None:
+        super().__init__(module)
+        self.chunk_manager = chunk_manager
+        self.param_op_hook = ZeROHookV2(chunk_manager)
+        self.fp32_params = []
+        # TODO: get param order and filter unused params
+        for p in module.parameters():
+            assert p.dtype == torch.half
+            fp32_p = p.float()
+            self.chunk_manager.append_tensor(p, 'fp16_param')
+            self.chunk_manager.append_tensor(fp32_p, 'fp32_param')
+            self.fp32_params.append(fp32_p)
+
+    def forward(self, *args, **kwargs):
+        self.module.zero_grad(set_to_none=True)
+        for p, fp32_p in zip(self.module.parameters(), self.fp32_params):
+            if not self.chunk_manager.is_chunk_free(p):
+                self.chunk_manager.update_tensor(p, fp32_p)
+        with use_param_op_hooks(self.param_op_hook):
+            outputs = self.module(*args, **kwargs)
+        self.chunk_manager.exec_lazy_release()
+        return outputs
+
+    def backward(self, loss: torch.Tensor):
+        with self.param_op_hook.switch_to_backward(), use_param_op_hooks(self.param_op_hook):
+            loss.backward()
+        self.chunk_manager.exec_lazy_release()
+        for p in self.module.parameters():
+            if self.chunk_manager.is_chunk_free(p) or not p.requires_grad:
+                p.grad = None
+            else:
+                p.grad = p.data
```

Comment:
what does it mean?
reusing grad fp16 with param fp16?
should it be p.data = p.grad?

Reply:
Set p.grad to correct ptr here. p.data saves grad, due to reuse.

Reply:
if a chunk is moved from gpu to cpu later. This line makes p.grad point to an old memory space.

Reply:
No, move device will update p.data at the same time.

Reply:
This line makes p.grad -> p.data (addr1)
Afterwards, you move chunk
let p.data (addr2)
However, p.grad still points to addr1.

Reply:
If so, just set p.grad again, or move this code snippets to `optimizer.step()` after device moving is done.

Reply:
To check grad in unit test, I just set `p.grad` here.

Reply:
Is it necessary to build a dict {param: chunk slice} to index grad and its true memory space (may reuse with param.data} ?

Reply:
Not necessary now I think. p.grad should always point to the chunk slice memory of p, as we always reuse now. If not reuse, I think it's necessary.

Reply:
OK, add it if necessary later.